c++/src/kj/compat/tls.c++

```diff
@@ -162,11 +163,23 @@ public:
       }
     });
   }
+
   kj::Promise<void> accept() {
     // We are the server. Set SSL options to prefer server's cipher choice.
     SSL_set_options(ssl, SSL_OP_CIPHER_SERVER_PREFERENCE);
 
-    return sslCall([this]() { return SSL_accept(ssl); }).ignoreResult();
+    auto sslAccept = [this]() {
+      return SSL_accept(ssl);
+    };
+
+    auto checkSslResult = [](size_t ret) {
+      if (ret == 0) {
+        kj::throwRecoverableException(
+            KJ_EXCEPTION(DISCONNECTED, "Client disconnected during SSL_accept()"));
+      }
+    };
+
+    return sslCall(kj::mv(sslAccept)).then(checkSslResult);
```

Comment:
Since you don't have `kj::mv()` on `checkSslResult`, I think this is capturing an lvalue reference, which is left dangling when `accept()` returns. KJ has a general convention that when a callback is expected, if you specify an lvalue, it is assumed you wanted to call back to that specific value by reference. I'll admit this convention is not my finest work -- in practice, almost no code ever actually intends to pass a callback by lvalue reference, and if it did, it would be clearer to wrap it in a lambda that explicitly captures the reference as a reference. Meanwhile, cases like this come up from time to time where someone accidentally passes a reference when they didn't intend to.

In this case, it doesn't result in any problem because the lambda doesn't actually capture anything, but if someone went and added a capture later, it would be corrupted.

I fear this problem will come up a lot more often if we're declaring callbacks on a separate line and then passing them into `.then()` as you've done here. Can we stick with the inline lambda style used throughout our code today?

Reply:
It is capturing an lvalue reference. I wouldn't say it's unintentional. I did encounter some segfaults from use-after-move in other places and chose to leave it unmoved when I could because it would be an easier skim. I agree about the capture situation. I'm uncomfortable with two points here:
- We're relying on an undocumented convention to prevent programmer error.
- We generally have made stateful functor usage unsafe.

I'll formulate it differently here to avoid that. Going forward, I'll try to see if I can either make then() have stronger guarantees or stand up a clang-tidy with bugprone-use-after-move.